File: src/SlimShader/Chunks/Rdef/ShaderType.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using SlimShader.Chunks.Common;
using SlimShader.Chunks.Shex;
using SlimShader.Util;

namespace SlimShader.Chunks.Rdef
{
	/// <summary>
	/// Describes a shader-variable type.
	/// Based on D3D11_SHADER_TYPE_DESC.
	/// </summary>
	public class ShaderType
	{
		private readonly int _indent;
		private readonly bool _isFirst;

		/// <summary>
		/// Identifies the variable class as one of scalar, vector, matrix or object.
		/// </summary>
		public ShaderVariableClass VariableClass { get; private set; }

		/// <summary>
		/// The variable type.
		/// </summary>
		public ShaderVariableType VariableType { get; private set; }

		/// <summary>
		/// Number of rows in a matrix. Otherwise a numeric type returns 1, any other type returns 0.
		/// </summary>
		public ushort Rows { get; private set; }

		/// <summary>
		/// Number of columns in a matrix. Otherwise a numeric type returns 1, any other type returns 0.
		/// </summary>
		public ushort Columns { get; private set; }

		/// <summary>
		/// Number of elements in an array; otherwise 0.
		/// </summary>
		public ushort ElementCount { get; private set; }

		public List<ShaderTypeMember> Members { get; private set; }

		public string BaseTypeName { get; private set; }

		public ShaderType(int indent, bool isFirst)
		{
			_indent = indent;
			_isFirst = isFirst;
			Members = new List<ShaderTypeMember>();
		}

		public static ShaderType Parse(BytecodeReader reader, BytecodeReader typeReader, ShaderVersion target,
			int indent, bool isFirst, uint parentOffset)
		{
			var result = new ShaderType(indent, isFirst)
			{
				VariableClass = (ShaderVariableClass) typeReader.ReadUInt16(),
				VariableType = (ShaderVariableType) typeReader.ReadUInt16(),
				Rows = typeReader.ReadUInt16(),
				Columns = typeReader.ReadUInt16(),
				ElementCount = typeReader.ReadUInt16()
			};

			var memberCount = typeReader.ReadUInt16();
			var memberOffset = typeReader.ReadUInt32();

			if (target.MajorVersion >= 5)
			{
				var parentTypeOffset = typeReader.ReadUInt32(); // Guessing
				var parentTypeReader = reader.CopyAtOffset((int) parentTypeOffset);
				var parentTypeClass = (ShaderVariableClass) parentTypeReader.ReadUInt16();
				var unknown4 = parentTypeReader.ReadUInt16();

				var unknown1 = typeReader.ReadUInt32();
				if (unknown1 != 0)
				{
					var unknownReader = reader.CopyAtOffset((int) unknown1);
					uint unknown5 = unknownReader.ReadUInt32();
				}

				var unknown2 = typeReader.ReadUInt32();
				var unknown3 = typeReader.ReadUInt32();

				var parentNameOffset = typeReader.ReadUInt32();
				var parentNameReader = reader.CopyAtOffset((int) parentNameOffset);
				result.BaseTypeName = parentNameReader.ReadString();
			}

			if (memberCount > 0)
			{
				var memberReader = reader.CopyAtOffset((int) memberOffset);
				for (int i = 0; i < memberCount; i++)
					result.Members.Add(ShaderTypeMember.Parse(reader, memberReader, target, indent + 4, i == 0,
						parentOffset));
			}

			return result;
		}

		public override string ToString()
		{
			var indentString = "// " + new string(Enumerable.Repeat(' ', _indent).ToArray());
			var sb = new StringBuilder();
			if (_isFirst)
				sb.AppendLine(indentString);
			switch (VariableClass)
			{
				case ShaderVariableClass.InterfacePointer:
				case ShaderVariableClass.MatrixColumns:
				case ShaderVariableClass.MatrixRows:
				{
					sb.Append(indentString);
					if (!string.IsNullOrEmpty(BaseTypeName)) // BaseTypeName is only populated in SM 5.0
					{
						sb.Append(string.Format("{0}{1}", VariableClass.GetDescription(), BaseTypeName));
					}
					else
					{
						sb.Append(VariableClass.GetDescription());
						sb.Append(VariableType.GetDescription());
						if (Columns > 1)
						{
							sb.Append(Columns);
							if (Rows > 1)
								sb.Append("x" + Rows);
						}
					}
					break;
				}
				case ShaderVariableClass.Vector:
				{
					sb.Append(indentString + VariableType.GetDescription());
					if (Columns > 1)
						sb.Append(Columns);
					break;
				}
				case ShaderVariableClass.Struct:
					{
						if (!_isFirst)
							sb.AppendLine(indentString);
						sb.AppendLine(indentString + "struct " + BaseTypeName);
						sb.AppendLine(indentString + "{");
						foreach (var member in Members)
							sb.AppendLine(member.ToString());
						sb.AppendLine("//");
						sb.Append(indentString + "}");
						break;
					}
				case ShaderVariableClass.Scalar:
					{
						sb.Append(indentString + VariableType.GetDescription());
						break;
					}
				default:
					throw new ArgumentOutOfRangeException("Variable class '" + VariableClass + "' is not currently supported.");
			}
			return sb.ToString();
		}
	}
}

File: src/cpp/Source/Chunks/Rdef/ShaderType.cpp
#include "PCH.h"
#include "ShaderType.h"

using namespace std;
using namespace SlimShader;

shared_ptr<ShaderType> ShaderType::Parse(const BytecodeReader& reader, 
										 BytecodeReader& typeReader, 
										 const ShaderVersion& target,
										 int indent, 
										 bool isFirst, 
										 uint32_t parentOffset)
{
	auto result = shared_ptr<ShaderType>(new ShaderType(indent, isFirst));
	result->_variableClass = static_cast<ShaderVariableClass>(typeReader.ReadUInt16());
	result->_variableType = static_cast<ShaderVariableType>(typeReader.ReadUInt16());
	result->_rows = typeReader.ReadUInt16();
	result->_columns = typeReader.ReadUInt16();
	result->_elementCount = typeReader.ReadUInt16();

	auto memberCount = typeReader.ReadUInt16();
	auto memberOffset = typeReader.ReadUInt32();

	if (target.GetMajorVersion() >= 5)
	{
		auto parentTypeOffset = typeReader.ReadUInt32(); // Guessing
		auto parentTypeReader = reader.CopyAtOffset(parentTypeOffset);
		auto parentTypeClass = static_cast<ShaderVariableClass>(parentTypeReader.ReadUInt16());
		auto unknown4 = parentTypeReader.ReadUInt16();

		auto unknown1 = typeReader.ReadUInt32();
		if (unknown1 != 0)
		{
			auto unknownReader = reader.CopyAtOffset(unknown1);
			uint32_t unknown5 = unknownReader.ReadUInt32();
		}

		auto unknown2 = typeReader.ReadUInt32();
		auto unknown3 = typeReader.ReadUInt32();

		auto parentNameOffset = typeReader.ReadUInt32();
		if (parentNameOffset > 0)
		{
			auto parentNameReader = reader.CopyAtOffset(parentNameOffset);
			result->_baseTypeName = parentNameReader.ReadString();
		}
	}

	if (memberCount > 0)
	{
		auto memberReader = reader.CopyAtOffset(memberOffset);
		for (int i = 0; i < memberCount; i++)
			result->_members.push_back(ShaderTypeMember::Parse(reader, memberReader, target, indent + 4, i == 0,
				parentOffset));
	}

	return result;
}

ShaderVariableClass ShaderType::GetVariableClass() const { return _variableClass; }
ShaderVariableType ShaderType::GetVariableType() const { return _variableType; }
uint16_t ShaderType::GetRows() const { return _rows; }
uint16_t ShaderType::GetColumns() const { return _columns; }
uint16_t ShaderType::GetElementCount() const { return _elementCount; }
const std::vector<ShaderTypeMember>& ShaderType::GetMembers() const { return _members; }
const std::string& ShaderType::GetBaseTypeName() const { return _baseTypeName; }

ShaderType::ShaderType(int indent, bool isFirst)
	: _indent(indent), _isFirst(isFirst)
{

}

std::ostream& SlimShader::operator<<(std::ostream& out, const ShaderType& value)
{
	auto indentString = "// " + string(value._indent, ' ');
	if (value._isFirst)
		out << indentString << endl;
	switch (value._variableClass)
	{
	case ShaderVariableClass::InterfacePointer:
	case ShaderVariableClass::MatrixColumns:
	case ShaderVariableClass::MatrixRows:
		{
			out << indentString;
			if (value._baseTypeName.size() > 0) // BaseTypeName is only populated in SM 5.0
			{
				out << ToString(value._variableClass) << value._baseTypeName;
			}
			else
			{
				out << ToString(value._variableClass);
				out << ToString(value._variableType);
				if (value._columns > 1)
				{
					out << value._columns;
					if (value._rows > 1)
						out << "x" << value._rows;
				}
			}
			break;
		}
	case ShaderVariableClass::Vector:
		{
			out << indentString << ToString(value._variableType);
			if (value._columns > 1)
				out << value._columns;
			break;
		}
	case ShaderVariableClass::Struct:
		{
			if (!value._isFirst)
				out << indentString << endl;
			out << indentString << "struct " << value._baseTypeName << endl;
			out << indentString << "{" << endl;
			for (auto& member : value._members)
				out << member << endl;
			out << "//" << endl;
			out << indentString << "}";
			break;
		}
	case ShaderVariableClass::Scalar:
		{
			out << indentString << ToString(value._variableType);
			break;
		}
	default:
		throw new runtime_error("Variable class '" + to_string((int) value._variableClass) + "' is not currently supported.");
	}
	return out;
}